devito/types/tensor.py

```diff
@@ -110,7 +111,19 @@ def __subfunc_setup__(cls, *args, **kwargs):
                 kwargs["name"] = "%s_%s%s" % (name, d.name, dims[j].name)
                 kwargs["staggered"] = (stagg[i][j] if stagg is not None
                                        else (NODE if i == j else (d, dims[j])))
-                funcs2[j] = cls._sub_type(**kwargs)
+                # Setup kwargs for subfunction
+                # Through rebuilding or user input, the kwargs could be
+                # Tensors as well from a per-component property
+                sub_kwargs = {}
+                for k, v in kwargs.items():
+                    if isinstance(v, MatrixBase):
+                        sub_kwargs[k] = v[i, j]
+                    else:
```

Comment:
it's imho one of those cases in which `elif isinstance(v, <what's needed here>): .... else: ....` would read more smoothly than a nested try-except (but I'm nitpicking here)